src/katgpucbf/utils.py

```diff
@@ -53,3 +58,48 @@ def parse_source(value: str) -> list[tuple[str, int]] | str:
         return [(ep.host, ep.port) for ep in endpoints]
     except ValueError:
         return value
+
+
+class DeviceStatus(Enum):
+    """Discrete `device-status` readings.
+
+    Doesn't follow the convention of :class:`aiokatcp.Sensor.Status` because
+    it's useful for `UNKNOWN` to have the highest (or worst) value.
+    """
+
+    OK = 1
+    DEGRADED = 2
+    FAIL = 3
+    UNKNOWN = 4
+
+
+class DeviceStatusSensor(aiokatcp.AggregateSensor):
+    """Summary sensor for quickly ascertaining device status.
+
+    This takes its value from the worst status of its target set of sensors, so
+    it's quick to identify if there's something wrong, or if everything is good.
+    """
+
+    def __init__(self, target: aiokatcp.SensorSet):
+        super().__init__(target=target, sensor_type=DeviceStatus, name="device-status")
+
+    def update_aggregate(
+        self,
+        updated_sensor: Optional[aiokatcp.Sensor[_T]],
+        reading: Optional[aiokatcp.Reading[_T]],
+        old_reading: Optional[aiokatcp.Reading[_T]],
+    ) -> aiokatcp.Reading[DeviceStatus]:  # noqa: D102
+        # For device status it's far simpler just to re-calculate everything
+        # each time, than to try and maintain state.
+        worst_status: aiokatcp.Sensor.Status = aiokatcp.Sensor.Status.UNKNOWN
+        for sensor in self.target.values():
+            if self.filter_aggregate(sensor):
+                worst_status = max(worst_status, sensor.status)
+
+        if worst_status == aiokatcp.Sensor.Status.NOMINAL:
+            return aiokatcp.Reading(sensor.timestamp, aiokatcp.Sensor.Status.NOMINAL, DeviceStatus.OK)
+        if worst_status == aiokatcp.Sensor.Status.WARN:
+            return aiokatcp.Reading(sensor.timestamp, aiokatcp.Sensor.Status.WARN, DeviceStatus.DEGRADED)
+        if worst_status == aiokatcp.Sensor.Status.UNKNOWN:  # Do we need to take this situation into account?
+            return aiokatcp.Reading(sensor.timestamp, aiokatcp.Sensor.Status.UNKNOWN, DeviceStatus.UNKNOWN)
```

Comment:
I would've thought yes, but the question still stands. Maybe Bruce has a better idea.

Reply:
That being said, I see why you've done the mapping this way, but it does seem odd that this doesn't actively account for an `aiokatcp.Sensor.Status.FAILURE`.
Is the 'behaviour' of a `device-status` sensor in an ICD somewhere?

Reply:
> Is the 'behaviour' of a `device-status` sensor in an ICD somewhere?

Yes, it's in the [katcp spec](https://katcp-python.readthedocs.io/en/latest/_downloads/361189acb383a294be20d6c10c257cb4/NRF-KAT7-6.0-IFCE-002-Rev5-1.pdf), appendix B.2, although it's largely left up to the device to decide what it means. It doesn't allow for the status to be unknown though.

I don't think this case can realistically happen within katgpucbf because it would require all sensors to be UNKNOWN, and some of them are permanently NOMINAL. But more generally if it does happen my feeling is one should assume the device is OK until there is evidence to the contrary.